### PR TITLE
fix TF compilation issues; fix version of TF to avoid future issues

### PR DESCRIPTION
**Pull request by hainan-xv**

### CODE REVIEW

File: tools/extras/install_tensorflow_cc.sh [hunk 1]
```diff
@@ -35,15 +36,16 @@ cd ../
 # now bazel is built
 git clone https://github.com/tensorflow/tensorflow
 cd tensorflow
+git reset --hard 49059695babc08df22f31124480142e7f6aec5eb
```
Comment (galv) [inline tools/extras/install_tensorflow_cc.sh]: I highly recommend `git checkout 49059695babc08df22f31124480142e7f6aec5eb` instead. It's non-destructive and more intuitive to seasoned git users.

I notice this version is particularly cutting edge. Any particular reason for this as opposed to a stable version like v1.4.0?

File: tools/extras/install_tensorflow_cc.sh [hunk 2]
```diff
@@ -3,6 +3,7 @@
 set -e
 
 #export JAVA_HOME=/LOCATION_ON_YOUR_MACHINE/java/jdk1.8.0_121
+export JAVA_HOME=/export/b02/hxu/TensorFlow/java/jdk1.8.0_121
```
Comment (galv) [inline tools/extras/install_tensorflow_cc.sh]: I don't think this is a great idea. You should just check that JAVA_HOME isn't empty string instead, with `[ ! -z $JAVA_HOME ]`

**Comment by akshayc11** [comment]
An Additional suggestion I would like to make is, since we are already compiling the tensorflow cc from source, why not also compile the python pip wheel from source too? 
The commands I used to build the wheel:

```
# After you have finished extras/build_tensorflow_cc.sh
cd ${KALDI_ROOT}/tools/tensorflow
export HOME=${KALDI_ROOT}/tools/tensorflow_build # should already exist, but you can store the wheel anywhere.
bazel build -c opt --config=cuda //tensorflow/tools/pip_package:build_pip_package
bazel-bin/tensorflow/tools/pip_package/build_pip_package ${HOME}/tensorflow_pkg
wheel_file=`find ${HOME}/tensorflow_pkg -iname *.whl`
pip install --user ${wheel_file}
```

This would be more general instructions compared to the ones currently present in the extras/install_tensorflow_py.sh

**Comment by hainan-xv** [comment]
Hmm, I see that it's more general this way, however I am not sure if it's a good idea to introduce dependencies between the python install script and cc install script. I'm going to push the recipes for n-best rescoring in the next weeks or so which does not require the use of the cc part, so people only need to install TF in the python to make it work.

**Comment by akshayc11** [comment]
True, but those instructions seem to work only for specific distributions for which the wheel has been built, and for highly specific versions of Cuda and CUDNN. In my case, for my dev machine, I am running it in centOS7 with cuda-9.0 and cudnn 7 (I got the cc code working after adding some dynamic library links that tensorflow was looking for). The above instructions generate a wheel package that is guaranteed to run on the system as long as the cc compilation was a success, with the same configuration that the cc was built on.
However, it's your choice

**Comment by ZhaoZeqing** [comment]
I have an error when run the script 'egs/ami/s5/local/tfrnnlm/run_lstm_fast.sh', it uses tf_py to get a rnnlm, and uses tf_cc to rescore, I got the error in rescoring.
I guess this may related to tensorflow's version. I installed the newest tf_py 1.11.0rc0, and installed tf_cc by cloning from github last week.

**The error is following:**

2018-09-18 09:17:49.399889: E tensorflow/core/common_runtime/executor.cc:643] Executor failed to create kernel. Invalid argument: NodeDef mentions attr 'index_type' not in Op<name=Fill; signature=dims:int32, value:T -> output:T; attr=T:type>; NodeDef: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const). (Check whether your GraphDef-interpreting binary is up to date with your GraphDef-generating binary.).
         [[Node: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const)]]
2018-09-18 09:17:49.400435: E tensorflow/core/common_runtime/executor.cc:643] Executor failed to create kernel. Invalid argument: NodeDef mentions attr 'index_type' not in Op<name=Fill; signature=dims:int32, value:T -> output:T; attr=T:type>; NodeDef: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const). (Check whether your GraphDef-interpreting binary is up to date with your GraphDef-generating binary.).
         [[Node: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const)]]
ERROR (lattice-lmrescore-tf-rnnlm-pruned[5.5.13~1-a0bc1]:AcquireInitialTensors():tensorflow-rnnlm.cc:188) Invalid argument: NodeDef mentions attr 'index_type' not in Op<name=Fill; signature=dims:int32, value:T -> output:T; attr=T:type>; NodeDef: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const). (Check whether your GraphDef-interpreting binary is up to date with your GraphDef-generating binary.).
         [[Node: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const)]]

[ Stack-Trace: ]
kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::FatalMessageLogger::~FatalMessageLogger()
kaldi::tf_rnnlm::KaldiTfRnnlmWrapper::AcquireInitialTensors()
kaldi::tf_rnnlm::KaldiTfRnnlmWrapper::KaldiTfRnnlmWrapper(kaldi::tf_rnnlm::KaldiTfRnnlmWrapperOpts const&, std::string const&, std::string const&, std::string const&, std::string const&)
main
__libc_start_main
lattice-lmrescore-tf-rnnlm-pruned() [0x417e59]

ERROR (lattice-lmrescore-tf-rnnlm-pruned[5.5.13~1-a0bc1]:AcquireInitialTensors():tensorflow-rnnlm.cc:188) Invalid argument: NodeDef mentions attr 'index_type' not in Op<name=Fill; signature=dims:int32, value:T -> output:T; attr=T:type>; NodeDef: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const). (Check whether your GraphDef-interpreting binary is up to date with your GraphDef-generating binary.).
         [[Node: Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros = Fill[T=DT_FLOAT, _output_shapes=[[1,200]], index_type=DT_INT32, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/concat, Train/Model/MultiRNNCellZeroState_1/BasicLSTMCellZeroState/zeros/Const)]]

**Comment by danpovey** [comment]
That looks like it would be hard to debug.  I recommend to use the Kaldi-native RNNLM setup-- it doesn't have these types of problems and the results are also a little better.

**Comment by ZhaoZeqing** [comment]
@danpovey It worked! But I also want to ask a question, will kaldi consider using a higher version of tensorflow to run tfrnnlm?  Thanks!

**Comment by danpovey** [comment]
I think we previously fixed the version in the script to avoid problems
that were happening due to changes in TensorFlow.  But if you find that a
more recent version works, I'm OK in principle to change the installation
script to use that.  I don't think any of us are working very hard on this
setup right now but I'm willing to merge a pull request if you have one
that fixes an issue with it.

Dan

**Pull request merged**